File: Java/Lesson087_SubsetsIi.java
// =============================================================
// MIT License | @analyticswithharry2026
// GitHub  : https://github.com/analyticswithharry
// YouTube : Analytics with Harry
// =============================================================
// Lesson     : 087 -- Subsets II
// Category   : Backtracking
// Difficulty : Medium
// Study Plan : Day 44
// =============================================================
//
// QUESTION:
//   Given an integer array nums that may contain duplicates, return all possible subsets (the power set), without duplicates.
//   Example: [1,2,2] -> [[],[1],[1,2],[1,2,2],[2],[2,2]].
// =============================================================

import java.util.*;
public class Lesson087_SubsetsIi {
    List<List<Integer>> res = new ArrayList<>();
    public List<List<Integer>> subsetsWithDup(int[] nums){
        Arrays.sort(nums); bt(nums,0,new ArrayList<>()); return res;
    }
    void bt(int[] nums, int i, List<Integer> cur){
        res.add(new ArrayList<>(cur));
        for (int j=i;j<nums.length;j++){
            if (j>i && nums[j]==nums[j-1]) continue;
            cur.add(nums[j]); bt(nums,j+1,cur); cur.remove(cur.size()-1);
        }
    }
    public static void main(String[] a){
        System.out.println(new Lesson087_SubsetsIi().subsetsWithDup(new int[]{1,2,2}));
    }
}

File: Go/Lesson_087_subsets_ii.go
//go:build ignore

// =============================================================
// MIT License | @analyticswithharry2026
// GitHub  : https://github.com/analyticswithharry
// YouTube : Analytics with Harry
// =============================================================
// Lesson     : 087 -- Subsets II
// Category   : Backtracking
// Difficulty : Medium
// Study Plan : Day 44
// =============================================================
//
// QUESTION:
//   Given an integer array nums that may contain duplicates, return all possible subsets (the power set), without duplicates.
//   Example: [1,2,2] -> [[],[1],[1,2],[1,2,2],[2],[2,2]].
// =============================================================

package main
import ( "fmt"; "sort" )
func subsetsWithDup(nums []int) [][]int {
    sort.Ints(nums); var res [][]int; cur := []int{}
    var bt func(int)
    bt = func(i int){
        tmp := make([]int, len(cur)); copy(tmp, cur); res = append(res, tmp)
        for j := i; j < len(nums); j++ {
            if j > i && nums[j] == nums[j-1] { continue }
            cur = append(cur, nums[j]); bt(j+1); cur = cur[:len(cur)-1]
        }
    }
    bt(0); return res
}
func main(){ fmt.Println(subsetsWithDup([]int{1,2,2})) }
